parsing/parser.mly

```diff
@@ -3392,44 +3392,75 @@ tuple_type:
    - applications of type constructors:   int, int list, int option list
    - variant types:                       [`A]
  *)
+
+
+(*
+  Delimited types:
+    - parenthesised type          (type)
+    - first-class module types    (module S)
+    - object types                < x: t; ... >
+    - variant types               [ `A ]
+    - extension                   [%foo ...]
+*)
+delimited_type_without_object_type:
```

Comment:
Note: I am okay with the current state, but I would actually slightly prefer to:

1. Call this one `delimited_type_supporting_local_open`
2. Moving the `Ptyp_extension` out of this group, so that we stop supporting `M.[%foo]` as a local open. I see no strong need for that syntax, and I can see hypothetical situations where we would want to define a different meaning for quantified expansions. (I'm not sure, but we regularly get bitten by having accepted too many things that we now want to give a different meaning to, so I would not support a syntax unless there is a clear need.)